XenAdmin/Wizards/PatchingWizard/PlanActions/PatchPrecheckOnHostPlanAction.cs

```diff
@@ -120,5 +125,29 @@ protected override void RunWithSession(ref Session session)
                 livePatchAttempts[host.uuid].Add(xenServerPatch.Uuid);
             }
         }
+
+        public static void RefreshUpdate(Host host, HostUpdateMapping mapping, Session session)
+        {
+            // re-introduce pool_update if needed
+            if (mapping is PoolUpdateMapping poolUpdateMapping 
+                && session.Connection.Cache.Pool_updates.FirstOrDefault(u => string.Equals(u.uuid, poolUpdateMapping.Pool_update.uuid, StringComparison.OrdinalIgnoreCase)) == null)
+            {
+                log.InfoFormat("Re-introduce update on `{0}`. Update = `{1}` (uuid = `{2}`; old opaque_ref = `{3}`)",
+                    host.Name(), poolUpdateMapping.Pool_update.Name(), poolUpdateMapping.Pool_update.uuid,
+                    poolUpdateMapping.Pool_update.opaque_ref);
+                try
+                {
+                    var newUpdateRef = Pool_update.introduce(session, poolUpdateMapping.Pool_update.vdi.opaque_ref);
+                    session.Connection.WaitForCache(newUpdateRef);
+                }
+                catch (Exception e)
+                {
+                    log.Error("Failed to introduce update", e);

```

Comment:
I'd log this as "re-introduce" for clarity.
Shouldn't we rethrow the exception to prevent it from continuing unnecessarily since the precheck will eventually fail?

Reply:
Yes, that's a good idea, and we can show a meaningfull message.